### PR TITLE
add clad (a clang plugin)

### PR DESCRIPTION
**Pull request by Vipul-Cariappa**
Adds the ability to build llvm external projects along with clang/llvm.

This is useful in our case where we want to ship clad, a plugin for clang, to compiler-explorer.

---

Clad is a source-transformation automatic differentiation (AD) library for C++.
Source at https://github.com/vgvassilev/clad

cc: @vgvassilev

### CODE REVIEW

**Comment by mattgodbolt** [comment]
The way I'm reading this is that every clang 18 will build `clad` with it. I don't understand the implications of this: does the external project build separately or as part of the compiler? It appears the latter, which means as well as merging this we would have to rebuild all the clang 18.* compilers to pick this up.

Does having `clad` in the compiler build affect any non-`clad` code generation? I'm concerned about this as the clang folks rely on us having well-known builds of the the compiler to report bugs etc.

**Comment by Vipul-Cariappa** [comment]
`clad` can be built separately as well, or with `clang` as an external project to `llvm`.
As a `clang` plugin, `clad` needs to be built with every different version of `clang`. We currently do not support `clang 19`. So we are sticking to `clang 18` for now.

`clad` does not affect `clang` in any way. `clang` is a dependency of `clad`, and is only used as a dependency.
The build results in the addition of `clad.so` shared library and necessary header files copied to the staging directory.

Hope this addresses your issues.

**Comment by vgvassilev** [comment]
> The way I'm reading this is that every clang 18 will build `clad` with it. I don't understand the implications of this: does the external project build separately or as part of the compiler? It appears the latter, which means as well as merging this we would have to rebuild all the clang 18.* compilers to pick this up.
> 
> Does having `clad` in the compiler build affect any non-`clad` code generation? I'm concerned about this as the clang folks rely on us having well-known builds of the the compiler to report bugs etc.

Hi @mattgodbolt, just to add to @Vipul-Cariappa's comment. Clad.so needs to be built against a particular clang version due to the fact that LLVM does not offer API/ABI stable interfaces. Then it is used only to do `clang -fplugin=clad.so`. Not sure what this PR exactly does, but I'd think we need to build `clad.so` against any `clang-18` build for example, and somehow have a separate combobox option in the interface which automatically uses `clang-18 -fplugin=clad.so -I/path/to/clad/include`...

**Comment by mattgodbolt** [comment]
Got it: so unless the user passes `-fplugin=clad.so` then the built clang is the same?

File: build/build.sh [hunk 1]
```diff
@@ -377,6 +386,16 @@ for PATCH_TO_APPLY in "${PATCHES_TO_APPLY[@]}"; do
     git -C "${ROOT}/llvm-project" apply "${PATCH_TO_APPLY}" -v
 done
 
+# Setup external projects
+mkdir -p ${EXTERNAL_PROJECT_DIR}
```
Comment (mattgodbolt) [inline build/build.sh:390]: Can you quote this please

File: build/build.sh [hunk 2]
```diff
@@ -4,6 +4,7 @@ set -exo pipefail
 
 ROOT=$PWD
 VERSION=$1
+EXTERNAL_PROJECT_DIR=${ROOT}/external
```
Comment (mattgodbolt) [inline build/build.sh:7]: do you mind moving this to above line 21? Just seems to be more relevant there

File: build/build.sh [hunk 3]
```diff
@@ -377,6 +386,16 @@ for PATCH_TO_APPLY in "${PATCHES_TO_APPLY[@]}"; do
     git -C "${ROOT}/llvm-project" apply "${PATCH_TO_APPLY}" -v
 done
 
+# Setup external projects
+mkdir -p ${EXTERNAL_PROJECT_DIR}
+for I in "${!EXTERNAL_PROJECT_URLS[@]}"; do
```
Comment (mattgodbolt) [inline build/build.sh:391]: can you use `EXTERNAL_PROJECT_URL` instead of `I` please, and similar below. thanks!

**Review by mattgodbolt — COMMENTED** [review]
Thanks for your info: some nitpick only :)

**Review by mattgodbolt — APPROVED** [review]
Thanks!

**Comment by mattgodbolt** [comment]
I'll manually kick off a clang 18.1.0 build and then we need to forcibly reinstall it (which is a bit of a manual process; we don't expect our compiler binaries for stable compilers to change so we have extremely heavy caching)

**Comment by Vipul-Cariappa** [comment]
Hi @mattgodbolt, I have opened (updated) PR at https://github.com/compiler-explorer/compiler-explorer/pull/7109 to add Clad as a compiler. Adding Clad as a compiler made the most sense in our case. If we add it as a library, the user would be responsible for selecting the correct version of clang to work with Clad.

**Pull request merged**